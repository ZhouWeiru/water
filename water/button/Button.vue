<template>
  <button :type="htmlType" :disabled="disabled" :class="[
    name,
    {
      [`${prefix}primary`]: type === 'primary',
      [`${prefix}dashed`]: type === 'dashed',
      [`${prefix}danger`]: type === 'danger',
      [`${prefix}ghost`]: ghost,
      [`${prefix}lg`]: size === 'large',
      [`${prefix}sm`]: size === 'small', [`${prefix}lg`]: size === 'large', [`${prefix}loading`]: loading, [`${prefix}circle`]: circle,
      [`${prefix}click`]: isClick,
      [`${prefix}disabled`]: disabled,
    },
  ]" @animationend="removeClickName" @click="clickFn($event)">
    <w-icon
      v-show="loading || icon"
      v-if="loading || icon"
      :type="icon ? icon : 'loading1'"
      :spin="loading || icon.indexOf('loading') > -1"
    />
    <span :class="`${prefix}text`" v-if="$slots.default">
      <slot></slot>
    </span>
  </button>
</template>

<script>
const name = 'w-button';
const prefix = `${name}-`;

export default {
  name: 'WButton',
  props: {
    type: {
      type: String,
      default: 'default',
    },
    size: String,
    icon: String,
    loading: Boolean,
    circle: Boolean,
    ghost: Boolean,
    disabled: Boolean,
    htmlType: {
      type: String,
      default: 'button',
    },
    click: {
      type: Function,
      default: () => {},
    },
  },
  data() {
    return {
      name,
      prefix,
      isClick: false,
    };
  },
  methods: {
    clickFn(evente) {
      this.isClick = true;
      this.click(evente);
      this.$emit('click', evente);
    },
    removeClickName() {
      this.isClick = false;
    },
  },
};
</script>
