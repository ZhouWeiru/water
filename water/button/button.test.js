import { mount } from 'vue-test-utils';
import Button from './Button';

describe('Button.vue', () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = mount(Button, {
      slots: {
        default: 'default',
      },
    });
  });

  it('检测快照是否一样。', (done) => {
    wrapper.vm.$nextTick(() => {
      try {
        expect(wrapper.vm.$el).toMatchSnapshot();
        done();
      } catch (err) {
        done.fail(err);
      }
    });
  });

  it('click 事件是否触发', (done) => {
    try {
      wrapper.update();
      const button = wrapper.find('.w-button');
      button.trigger('click');
      expect(wrapper.vm.isClick).toBeTruthy();
      button.trigger('animationend');
      expect(wrapper.vm.isClick).toBeFalsy();
      done();
    } catch (err) {
      done.fail(err);
    }
  });
});
