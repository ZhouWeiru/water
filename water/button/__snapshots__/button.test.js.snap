// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue 检测快照是否一样。 1`] = `
<button
  class="w-button"
  type="button"
>
  <!---->
   
  <span
    class="w-button-text"
  >
    default
  </span>
</button>
`;
